synapse/rest/client/room.py

```diff
@@ -929,48 +1004,58 @@ async def on_POST(self, request):
 class JoinedRoomsRestServlet(RestServlet):
     PATTERNS = client_patterns("/joined_rooms$", v1=True)
 
-    def __init__(self, hs):
+    def __init__(self, hs: "HomeServer"):
         super().__init__()
         self.store = hs.get_datastore()
         self.auth = hs.get_auth()
 
-    async def on_GET(self, request):
+    async def on_GET(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
         requester = await self.auth.get_user_by_req(request, allow_guest=True)
 
         room_ids = await self.store.get_rooms_for_user(requester.user.to_string())
         return 200, {"joined_rooms": list(room_ids)}
 
 
-def register_txn_path(servlet, regex_string, http_server, with_get=False):
+def register_txn_path(
+    servlet: RestServlet,
+    regex_string: str,
+    http_server: HttpServer,
+    with_get: bool = False,
+) -> None:
     """Registers a transaction-based path.
 
     This registers two paths:
         PUT regex_string/$txnid
         POST regex_string
 
     Args:
-        regex_string (str): The regex string to register. Must NOT have a
-        trailing $ as this string will be appended to.
-        http_server : The http_server to register paths with.
+        regex_string: The regex string to register. Must NOT have a
+            trailing $ as this string will be appended to.
+        http_server: The http_server to register paths with.
         with_get: True to also register respective GET paths for the PUTs.
     """
-    http_server.register_paths(
-        "POST",
-        client_patterns(regex_string + "$", v1=True),
-        servlet.on_POST,
-        servlet.__class__.__name__,
-    )
-    http_server.register_paths(
-        "PUT",
-        client_patterns(regex_string + "/(?P<txn_id>[^/]*)$", v1=True),
-        servlet.on_PUT,
-        servlet.__class__.__name__,
-    )
-    if with_get:
+    on_POST = getattr(servlet, "on_POST", None)
+    if on_POST:
+        http_server.register_paths(
+            "POST",
+            client_patterns(regex_string + "$", v1=True),
+            on_POST,
+            servlet.__class__.__name__,
+        )
+    on_PUT = getattr(servlet, "on_PUT", None)
```

Comment:
Do we want to allow servlet classes to not have a `PUT` here? This feels more like we should be asserting here instead (which will cause Synapse not to start)

Reply:
That's fine. I figured we can go either way. Annoyingly an mypy doesn't seem to work great with `hasattr`. 😢 I'll take another look!